lib/puppet_pal.rb

```diff
@@ -23,26 +23,237 @@
+    end
+
+    # Evaluates a puppet language file in top scope.
+    # The file must exist and contain valid puppet language code or an error is raised.
+    #
+    # @param file [Path, String] an absolute path to a file with puppet language code, must exist
+    # @return [Object] what the last evaluated expression in the file evaluated to
+    #
+    def evaluate_file(file)
+      internal_evaluator.evaluate_file(topscope, file)
+    end
+
+    # Parses a puppet data type given in String format and returns that type, or raises an error.
+    # A type is needed in calls to `new` to create an instance of the data type, or to perform type checking
+    # of values - typically using `type.instance?(obj)` to check if `obj` is an instance of the type.
+    #
+    # @example Verify if obj is an instance of a data type
+    #   # evaluates to true
+    #   pal.type('Enum[red, blue]').instance?("blue")
+    #
+    # @example Create an instance of a data type
+    #   # using an already create type
+    #   t = pal.type('Car')
+    #   pal.new_object(t, [{'color' => 'black', 'make' => 't-ford}])
+    #
+    #   # letting 'new_object' parse the type from a string
+    #   pal.new_object('Car', [{'color' => 'black', 'make' => 't-ford}])
+    #
+    # @param type_string [String] a puppet language data type
+    # @return [Puppet::Pops::Types::TypedModelObject] the data type
+    #
+    def type(type_string)
+      Puppet::Pops::Types::TypeParser.singleton.parse(type_string)
+    end
+
+    # Creates a new instance of a given data type.
+    # @param data_type [String, Puppet::Pops::Types::PAnyType] the data type as a data type or in String form.
+    # @param arguments [Array] an array of the arguments to `new`, defaults to empty array.
+    # @return [Object] an instance of the given data type,
+    #   or raises an error if it was not possible to parse data type or create an instance.
+    #
+    def new_object(data_type, arguments = [])
+      Pal::assert_optionally_empty_array(arguments, 'arguments')
+      t = data_type.is_a?(String) ? type(data_type) : data_type
+      unless t.is_a?(Puppet::Pops::Types::PAnyType)
+        raise ArgumentError, _("Given data_type value is not a data type, got '%{type}'") % {type: t.class}
+      end
+      call_function('new', [t] + arguments)
+    end
+
+    private
+
+    def topscope
+      internal_compiler.topscope
+    end
+  end
+
+  class ScriptCompiler < Compiler
+    # Returns the signature Callable of the given plan (the arguments it accepts, and the data type it returns)
+    # @param plan_name [String] the name of the plan to get the signature of
+    # @return [Callable, nil] returns a Callable data type, or nil if plan is not found
+    #
+    def plan_signature(plan_name)
+      loader = internal_compiler.loaders.private_environment_loader
+      if loader && func = loader.load(:plan, plan_name)
+          require 'byebug'; debugger
```

Comment:
It doesn't appear this code path is being tested.

Reply:
:-) true